src/codeeditor/editor.ts

```diff
@@ -0,0 +1,637 @@
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    mimeTypeChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
+
+    /**
+     * Get the number of lines in the model.
+     */
+    readonly lineCount: number;
+
+    /**
+     * The currently selected code.
+     *
+     * @returns A read-only copy of the text selections.
+     */
+    readonly selections: ObservableVector<ITextSelection>;
+
+    /**
+     * Returns the primary cursor position.
+     */
+    getCursorPosition(): IPosition;
+
+    /**
+     * Returns the content for the given line number.
+     */
+    getLine(line: number): string;
+
+    /**
+     * Find an offset for the given position.
+     */
+    getOffsetAt(position: IPosition): number;
+
+    /**
+     * Find a position for the given offset.
+     */
+    getPositionAt(offset: number): IPosition;
+
+    /**
+     * Undo one edit (if any undo events are stored).
+     */
+    undo(): void;
+
+    /**
+     * Redo one undone edit.
+     */
+    redo(): void;
+
+    /**
+     * Clear the undo history.
+     */
+    clearHistory(): void;
+
+    /**
+     * Update mime type 
+     */
+    setMimeTypeFromPath(path: string): void;
+  }
+
+  /**
+   * A keydown handler type.
+   */
+  export
+  type KeydownHandler = (instance: IEditor, event: KeyboardEvent) => boolean;
+
+  /**
+   * A widget that provides a code editor.
+   */
+  export
+  interface IEditor extends IDisposable {
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The model used by the editor.
+     */
+    readonly model: IModel;
+
+    /**
+     * The height of a line in the editor in pixels.
+     */
+    readonly lineHeight: number;
+
+    /**
+     * The widget of a character in the editor in pixels.
+     */
+    readonly charWidth: number;
+
+    /**
+     * Handle keydown events for the editor.
+     *
+     * #### Notes
+     * Return `true` to prevent the default handling of the event by the
+     * editor.
+     */
+    onKeyDown: KeydownHandler;
+
+    /**
+     * Brings browser focus to this editor text.
+     */
+    focus(): void;
+
+    /**
+     * Repaint editor. 
+     */
+    refresh(): void;
+
+    /**
+     * Test whether the editor has keyboard focus.
+     */
+    hasFocus(): boolean;
+
+    /**
+     * Set the size of the editor in pixels.
+     */
+    setSize(width: number, height: number): void;
+
+    /**
+     * Scroll the given cursor position into view.
+     */
+    scrollIntoView(pos: IPosition, margin?: number): void;
```

Comment:
Do we need margin or it is impl details of CodeMirror? For Monaco there are methods like `revealPosition` and `revealSelection` which not accept margin. Also I like `reveal` better than `scrollIntoView`. Could we rename?